### PR TITLE
fix(apprunner-alpha): env vars and secrets can't solely be added via .add*() methods

### LINKED ISSUE #24345: (apprunner-alpha): only using .addEnvrionmentVariable() and .addSecret() fails to work
### Describe the bug

Using the apprunner-alpha module, methods `.addEnvironmentVariable()` and `.addSecret()` only work if there are existing entries that have been created in the initial `imageConfiguration` props.  This is because in the private `renderEnvironmentVariables()` and `renderEnvironmentSecrets()` methods, there is an if condition which returns undefined if there are no entries in the initial `imageConfiguration` prop. See code here https://github.com/aws/aws-cdk/blob/d986e143df3cf9b42031eba0f5a2d9a71d6d9208/packages/%40aws-cdk/aws-apprunner/lib/service.ts#L1139-L1167

### Expected Behavior

Ability to use the methods without having defined initial env variables in the props.

### Current Behavior

No environment or secret variables are rendered if only using the `.addEnvironmentVariable()` and `.addSecret()` methods.

### Reproduction Steps

```
    const service = new Service(this, 'service', {
      source: Source.fromAsset({
        imageConfiguration: {
          port: 3000,
        },
        asset: imageAsset,
      }),
      instanceRole,
    })
    
    service.addEnvironmentVariable('TEST', 'test')
    service.addSecret('SECRET', Secret.fromSecretsManager(secret)
```

This will not render any secrets or environment variables in the resulting template, but adding some initial values will cause add environment variables to be correctly rendered, e.g.

```
    const service = new Service(this, 'service', {
      source: Source.fromAsset({
        imageConfiguration: {
          port: 3000,
          environmentSecrets: { SEED1: Secret.fromSecretsManager(seedSecret) },
          environmentVariables: { SEED2: 'seed' },
        },
        asset: imageAsset,
      }),
      instanceRole,
    })
    
    service.addEnvironmentVariable('TEST', 'test')
    service.addSecret('SECRET', Secret.fromSecretsManager(secret)
```

### Possible Solution

Correct the faulty if logic in `renderEnvironmentVariables()` and `renderEnvironmentSecrets()`

### Additional Information/Context

_No response_

### CDK CLI Version

2.66.0

### Framework Version

_No response_

### Node.js Version

16.17.0

### OS

M1 MacOS

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by rogerchi**
This fixes the logic for rendering environment variables and environment secrets for the `apprunner-alpha` module.  Previously, `.addEnvironmentVariable()` and `.addSecret()` were being ignored if there were not already "seed" values in the input props.

Closes #24345.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by TheRealAmazonKendra** [comment]
Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))

**Comment by rogerchi** [comment]
> Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))

I've updated the code, tests, and added a new integration test and snapshot.

**Comment by TheRealAmazonKendra** [comment]
> > Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))
> 
> I've updated the code, tests, and added a new integration test and snapshot.

I'll take a look but please also address the message above.

**Comment by rogerchi** [comment]
> I'll take a look but please also address the message above.

Let me know if the new title works.

File: packages/@aws-cdk/aws-apprunner/test/service.test.ts [hunk 1]
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
   });
 });
 
+
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/test/service.test.ts:394]:
```suggestion
```

File: packages/@aws-cdk/aws-apprunner/test/service.test.ts [hunk 2]
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
...
+      },
+      imageIdentifier: 'public.ecr.aws/aws-containers/hello-app-runner:latest',
+    }),
+  });
+
+  service.addEnvironmentVariable('TEST_ENVIRONMENT_VARIABLE', 'test environment variable value');
+  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
+
+  // THEN
+  // we should have the service
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/test/service.test.ts:414]:
```suggestion
```

File: packages/@aws-cdk/aws-apprunner/test/service.test.ts [hunk 3]
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
...
+    source: apprunner.Source.fromEcrPublic({
+      imageConfiguration: {
+        startCommand: '/root/start-command.sh',
+      },
+      imageIdentifier: 'public.ecr.aws/aws-containers/hello-app-runner:latest',
+    }),
+  });
+
+  service.addEnvironmentVariable('TEST_ENVIRONMENT_VARIABLE', 'test environment variable value');
+  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/test/service.test.ts:411]:
```suggestion
  const secret = new secretsmanager.Secret(stack, 'Secret');
  const service = new apprunner.Service(stack, 'DemoService', {
    source: apprunner.Source.fromEcrPublic({
      imageConfiguration: {
        startCommand: '/root/start-command.sh',
      },
      imageIdentifier: 'public.ecr.aws/aws-containers/hello-app-runner:latest',
    }),
  });
  
  // WHEN
  service.addEnvironmentVariable('TEST_ENVIRONMENT_VARIABLE', 'test environment variable value');
  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
```

File: packages/@aws-cdk/aws-apprunner/test/service.test.ts [hunk 4]
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
   });
 });
 
+
+test('custom environment variables and secrets can be added without first defining them in props', () => {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/test/service.test.ts:395]: Can we split this into two tests, one for each add*() method? If someone breaks one of these methods but not another, having the name of the test match the method that was broken will make it easier to debug.

File: packages/@aws-cdk/aws-apprunner/lib/service.ts [hunk 5]
```diff
@@ -1130,36 +1132,22 @@ export class Service extends cdk.Resource {
...
   private renderEnvironmentSecrets(): EnvironmentSecret[] | undefined {
-    if (Object.keys(this.environmentSecrets).length > 0 && this.instanceRole) {
-      for (const [key, value] of Object.entries(this.environmentSecrets)) {
-        if (key.startsWith('AWSAPPRUNNER')) {
-          throw new Error(`Environment secret key ${key} with a prefix of AWSAPPRUNNER is not allowed`);
-        }
-
-        value.grantRead(this.instanceRole);
-        this.secrets.push({ name: key, value: value.arn });
-      }
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/lib/service.ts:1162]: Why was this removed?

File: packages/@aws-cdk/aws-apprunner/lib/service.ts [hunk 6]
```diff
@@ -1130,36 +1132,22 @@ export class Service extends cdk.Resource {
...
   }
 
   private renderEnvironmentVariables(): EnvironmentVariable[] | undefined {
-    if (Object.keys(this.environmentVariables).length > 0) {
-      for (const [key, value] of Object.entries(this.environmentVariables)) {
-        if (key.startsWith('AWSAPPRUNNER')) {
-          throw new Error(`Environment variable key ${key} with a prefix of AWSAPPRUNNER is not allowed`);
-        }
-        this.variables.push({ name: key, value: value });
-      }
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/lib/service.ts:1146]: Why was this removed?

File: packages/@aws-cdk/aws-apprunner/lib/service.ts [hunk 7]
```diff
@@ -981,17 +972,22 @@ export class Service extends cdk.Resource {
...
+    }
 
     // generate an IAM role only when ImageRepositoryType is ECR and props.accessRole is undefined
     this.accessRole = (this.source.imageRepository?.imageRepositoryType == ImageRepositoryType.ECR) ?
       this.props.accessRole ?? this.generateDefaultRole() : undefined;
 
-    // generalte an IAM role only when environmentSecrets has values and props.instanceRole is undefined
-    this.instanceRole = (Object.keys(this.environmentSecrets).length > 0 && !this.props.instanceRole) ?
-      this.createInstanceRole() : this.props.instanceRole;
-
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-apprunner/lib/service.ts:994]: This same logic now needs to be performed lazily, no? Not entirely removed

**Review by comcalvi — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-apprunner/test/service.test.ts (hunk 4 above)
Comment (rogerchi) [reply packages/@aws-cdk/aws-apprunner/test/service.test.ts]: Done

File: packages/@aws-cdk/aws-apprunner/lib/service.ts (hunk 7 above)
Comment (rogerchi) [reply packages/@aws-cdk/aws-apprunner/lib/service.ts:994]: This is handled in lines 975, 1049, and then lazily produced in line 1000.

File: packages/@aws-cdk/aws-apprunner/lib/service.ts (hunk 5 above)
Comment (rogerchi) [reply packages/@aws-cdk/aws-apprunner/lib/service.ts:1162]: The logic for throwing the error now happens in the .addSecret() method.  And the prop secrets are added using the same .addSecret() method now, see lines 977:985.  I followed the pattern that the `lambda.Function` construct established by using the .add* methods for both prop environment variables and ones added later using the .add* methods.

File: packages/@aws-cdk/aws-apprunner/lib/service.ts (hunk 6 above)
Comment (rogerchi) [reply packages/@aws-cdk/aws-apprunner/lib/service.ts:1146]: See comment above

**Comment by rogerchi** [comment]
@comcalvi I've tried to address/answer all of your comments.  Let me know if there's any further suggestions or clarifications required!

**Review by comcalvi — APPROVED** [review]
nice work!
